subsys/net/lib/lwm2m/lwm2m_observation.c

```diff
@@ -87,12 +87,142 @@ static const uint8_t LWM2M_ATTR_LEN[] = {4, 4, 2, 2, 2};
 
 static struct lwm2m_attr write_attr_pool[CONFIG_LWM2M_NUM_ATTR];
 
+/* Notification value register is used to store the resource values sent in
+ * the last notification, for resources that carry one of the numerical
+ * gt/lt/st attributes. This value is needed to determine whether notification
+ * send conditions are satisfied for given attributes.
+ */
+static struct lwm2m_notify_value_register {
+	/* A pointer to the corresponding resource instance. */
+	const struct lwm2m_engine_res_inst *ref;
+
+	/* Last notified resource value. */
+	double value;
+
+	/* Whether resource was notified already or not. */
+	bool notified : 1;
+} notify_value_pool[CONFIG_LWM2M_MAX_NOTIFIED_NUMERICAL_RES_TRACKED];
+
 /* Forward declarations */
 
 void lwm2m_engine_free_list(sys_slist_t *path_list, sys_slist_t *free_list);
 
 struct lwm2m_obj_path_list *lwm2m_engine_get_from_list(sys_slist_t *path_list);
 
+static struct lwm2m_notify_value_register *notify_value_reg_alloc(
+	const struct lwm2m_engine_res_inst *ref)
+{
+	for (int i = 0; i < ARRAY_SIZE(notify_value_pool); i++) {
+		if (notify_value_pool[i].ref == NULL) {
+			notify_value_pool[i].ref = ref;
+			notify_value_pool[i].notified = false;
+
+			return &notify_value_pool[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void notify_value_reg_free(const struct lwm2m_engine_res_inst *ref)
+{
+	for (int i = 0; i < ARRAY_SIZE(notify_value_pool); i++) {
+		if (ref == notify_value_pool[i].ref) {
+			(void)memset(&notify_value_pool[i], 0,
+				     sizeof(notify_value_pool[i]));
+			break;
+		}
+	}
+}
+
+static struct lwm2m_notify_value_register *notify_value_reg_get(
+	const struct lwm2m_engine_res_inst *ref)
+{
+	for (int i = 0; i < ARRAY_SIZE(notify_value_pool); i++) {
+		if (ref == notify_value_pool[i].ref) {
+			return &notify_value_pool[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int notify_value_reg_inst_update(const struct lwm2m_engine_res_inst *ref,
+					bool enable)
+{
+	struct lwm2m_notify_value_register *notify_reg;
+
+	if (ref->res_inst_id == RES_INSTANCE_NOT_CREATED) {
+		return 0;
+	}
+
+	if (!enable) {
+		notify_value_reg_free(ref);
+		return 0;
+	}
+
+	notify_reg = notify_value_reg_get(ref);
+	if (notify_reg == NULL) {
+		notify_reg = notify_value_reg_alloc(ref);
+	}
+
+	if (notify_reg == NULL) {
```

Comment:
Drop?

Reply:
I've ended up keeping the if as I've also added a log, fixed the return statement below.